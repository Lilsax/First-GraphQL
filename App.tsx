/* eslint-disable eqeqeq */
/* eslint-disable quotes */
/* eslint-disable semi */
/* eslint-disable react-native/no-inline-styles */
/* eslint-disable prettier/prettier */
/**
 * Sample React Native App
 * https://github.com/facebook/react-native
 *
 * @format
 */

import React from "react";
import Navigation from './navigation';

function App(): React.JSX.Element {
  return (
    < Navigation />
  );
}

export default App;
