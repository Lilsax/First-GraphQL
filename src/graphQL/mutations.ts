/* eslint-disable prettier/prettier */

import { gql } from '@apollo/client';

const increaseTrackView = gql`
mutation IncrementTrackViews($incrementTrackViewsId: ID!) {
  incrementTrackViews(id: $incrementTrackViewsId) {
    track {
      modules {
        id
        title
        length
        content
        videoUrl
      }
      author {
        id
        name
        photo
      }
      id
      title
      length
      numberOfViews
      description
      modulesCount
      thumbnail
    }
    message
    success
    code
  }
}`;

export { increaseTrackView };
