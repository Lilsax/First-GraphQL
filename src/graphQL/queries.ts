/* eslint-disable prettier/prettier */
import { gql } from '@apollo/client';

const getHomeData = gql`
    query getTracksHome {
        tracksForHome {
            id
            thumbnail
            numberOfViews
            title
            modules {
            id
            title
            length
            content
            videoUrl
            }
            length
            description
            author {
            id
            name
            photo
            }
            modulesCount
        }
    }
`;

const getCatInfo = gql`
    query getCatInfo($trackId: ID!) {
        track(id: $trackId) {
            title
            id
            modules {
            id
            title
            length
            content
            videoUrl
            }
            author {
            id
            name
            photo
            }
            numberOfViews
            description
            modulesCount
            length
            thumbnail
        }
}`

export { getHomeData, getCatInfo };
