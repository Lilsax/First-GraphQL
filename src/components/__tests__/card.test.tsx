/* eslint-disable quotes */
/* eslint-disable prettier/prettier */
import React from "react";
import Card from "../card";
import { render, fireEvent, screen } from "@testing-library/react-native";
import App from "../../../App";
import { getCatInfo, getHomeData } from "../../graphQL/queries";
import { MockedProvider } from "@apollo/client/testing";
import { increaseTrackView } from "../../graphQL/mutations";

describe("testing the full functionlity of the card", () => {
  const pressFunction = jest.fn();
  const btnTestID = "card-btn";
  const imgTestID = "image-test";
  const imageURLTest = "www.google.com/img";
  const cardItem = (
    <Card
      handlePress={pressFunction}
      thumbnail={imageURLTest}
      title="title test"
      numberOfViews={1}
      description="description test"
      btnTestID={btnTestID}
      imgTestID={imgTestID}
    />
  );

  const errorMock = [
    {
      request: {
        query: getHomeData,
      },
      error: new Error("An error occurred"),
    },
  ];

  const successMocks = [
    {
      request: {
        query: getHomeData,
      },
      result: {
        data: {
          tracksForHome: [
            {
              id: "c_0",
              title: "Cat-stronomy, an introduction",
              thumbnail:
                "https://res.cloudinary.com/dety84pbu/image/upload/v1598465568/nebula_cat_djkt9r.jpg",
              length: 2377,
              modulesCount: 10,
              numberOfViews: 8,
              description:
                "Curious to learn what Cat-stronomy is all about? Explore the planetary and celestial alignments and how they have affected our space missions.",
            },
          ],
        },
      },
    },
    {
      request: {
        query: getCatInfo,
        variables: { trackId: "c_0" },
      },
      result: {
        data: {
          track: {
            thumbnail:
              "https://res.cloudinary.com/dety84pbu/image/upload/v1598474100/famous_cats_epuqcr.jpg",
            description:
              "Be inspired by famous catstronauts who have made their names legend from across the galaxies. Special guest appearance included, so hold on to your boots!",
          },
        },
      },
    },
    {
      request: {
        query: increaseTrackView,
        variables: { incrementTrackViewsId: "c_0" },
      },
      result: {
        data: {
          incrementTrackViews: {
            track: {
              id: "c_0",
              title: "Cat-stronomy, an introduction",
              thumbnail:
                "https://res.cloudinary.com/dety84pbu/image/upload/v1598465568/nebula_cat_djkt9r.jpg",
              length: 2377,
              modulesCount: 10,
              numberOfViews: 8,
              description:
                "Curious to learn what Cat-stronomy is all about? Explore the planetary and celestial alignments and how they have affected our space missions.",
            },
          },
        },
      },
    },
  ];

  test("Snapshot test", () => {
    const { toJSON } = render(cardItem);
    expect(toJSON()).toMatchSnapshot();
  });
  test("title rendered correct", () => {
    const { getByText } = render(cardItem);
    expect(getByText("title test")).toBeTruthy();
  });

  test("image has the right src", () => {
    const { getByTestId } = render(cardItem);
    const thumbnailImage = getByTestId("image-test");
    expect(thumbnailImage.props.source.uri).toEqual(imageURLTest);
  });

  test("numberOfViews rendered correct", () => {
    const { getByText } = render(cardItem);
    expect(getByText("Number Of Views : 1")).toBeTruthy();
  });

  test("on button press function called ", () => {
    const { getByTestId } = render(cardItem);
    fireEvent.press(getByTestId("card-btn"));
    expect(pressFunction).toHaveBeenCalledTimes(1);
  });

  test("on button press it will navigate to another screen ", async () => {
    render(
      <MockedProvider mocks={successMocks} addTypename={false}>
        <App />
      </MockedProvider>
    );

    expect(await screen.findByText("<<<< LOADING >>>>")).toBeDefined();
    expect(
      await screen.findByText("Cat-stronomy, an introduction")
    ).toBeDefined();

    fireEvent.press(await screen.findByTestId("card-btn"));
    expect(await screen.findByTestId("detailsScreenView")).toBeTruthy();
  });

  test("show an error when graphQL fail ", async () => {
    render(
      <MockedProvider mocks={errorMock} addTypename={false}>
        <App />
      </MockedProvider>
    );
    expect(await screen.findByText("An error occurred")).toBeDefined();
  });


});
