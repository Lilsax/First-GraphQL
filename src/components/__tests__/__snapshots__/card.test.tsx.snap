// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`testing the full functionlity of the card Snapshot test 1`] = `
<View
  style={
    {
      "borderRadius": 5,
      "marginVertical": 20,
      "width": "100%",
    }
  }
>
  <View
    accessibilityState={
      {
        "busy": undefined,
        "checked": undefined,
        "disabled": undefined,
        "expanded": undefined,
        "selected": undefined,
      }
    }
    accessibilityValue={
      {
        "max": undefined,
        "min": undefined,
        "now": undefined,
        "text": undefined,
      }
    }
    accessible={true}
    collapsable={false}
    focusable={true}
    onClick={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      {
        "opacity": 1,
      }
    }
    testID="card-btn"
  >
    <Image
      alt="some image"
      source={
        {
          "uri": "www.google.com/img",
        }
      }
      style={
        {
          "borderRadius": 10,
          "height": 300,
          "marginBottom": 20,
          "objectFit": "cover",
          "width": "100%",
        }
      }
      testID="image-test"
    />
    <Text
      style={
        {
          "color": "black",
          "fontSize": 18,
          "fontStyle": "italic",
          "marginBottom": 5,
          "textAlign": "left",
        }
      }
    >
      title test
    </Text>
    <Text
      style={
        {
          "fontStyle": "italic",
          "textAlign": "left",
        }
      }
    >
      Number Of Views : 1
    </Text>
  </View>
</View>
`;
